### PR TITLE
Update Build Toolchain

### PR DESCRIPTION
**Pull request by thalassemia**
This removes the deprecated `distutils` in favor of `setuptools` and replaces the `setup.py` CLI with `pip` and `build`.

Here is a table of environments where `python stochastic_arrow/test/test_arrow.py` ran without issue after installing either an `sdist` (platform-agnostic) or `wheel` (platform-specific) that I built on macOS 15.1.1, Python 3.12, Cython 3.0.11, and Numpy 2.1.3 with `python -m build --sdist/wheel`:

| Operating System | Python version | Numpy version | Build type |
|------------------|----------------|---------------|----|
| Docker `debian:latest`           | 2.7            | 1.14.0        | `sdist` |
| Docker `debian:latest`           | 2.7            | 1.16.0        | `sdist` |
| macOS 15.1.1            | 3.12            | 1.26.4        | `sdist` |
| macOS 15.1.1            | 3.12            | 1.26.4        | `wheel` |

It's pretty cool that this can still be installed on Python 2.7 with just a few tweaks. The last row suggests that building with Numpy 2.1.3 does not make the wheel incompatible with older versions of Numpy.

Also, I added a `cibuildwheel` GitHub Action that should automatically build and publish wheels for all common platforms and recent Python versions. It should run `pytest` on every wheel it builds so we won't accidentally upload a non-functional wheel.

I got the wheel working on the new free-threaded release of Python 3.13. Unfortunately, it was almost universally slower than the normal Python 3.13 both to build and to run `pytest`.

### CODE REVIEW

**Comment by thalassemia** [comment]
In the process of getting `pytest` to work with `cibuildwheel`, I realized that our current directory structure can lead to import issues when trying to test a wheel that you built and installed. Specifically, calling `pytest` inside the root directory will import the `stochastic_arrow` source files instead of the installed wheel.

To fix this, I decided to follow [PyPA's recommendation](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/) to have all package source files inside a `src` directory. I could then move the `test` directory out to the top level and move the `data` directory into `test`. From there, I just had to modify a few tests to load data from the right place. Now, `pytest` works seamlessly regardless of whether it is using an editable install ("Pytest" GitHub Action) or an installed wheel ("Build and upload to PyPI" GitHub Action).

Also, I added a temporary job to `release.yml` to upload the built wheels to TestPyPI, a mock PyPI. You can view [the results of that here](https://test.pypi.org/project/stochastic-arrow/) and try installing using `pip install -i https://test.pypi.org/simple/ stochastic-arrow`.

File: src/stochastic_arrow/arrow.pxd [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # cython: language_level=3str
+# cython: freethreading_compatible = True
```
Comment (1fish2) [inline src/stochastic_arrow/arrow.pxd:2]: Does it suffice to declare this in the .pxy module?

**Review by 1fish2 — APPROVED** [review]
This looks great!

I fixed the CLI test runner although bugs remain in the `--plot` and `--memory` tests. See the commit message.

**Comment by thalassemia** [comment]
Thanks for the review and edits @1fish2! I noticed that the memory test in its current form fails on MacOS and changed it to `assert memory_increases <= 10` instead of `<= 1`. Not sure why though, or if it's even worth digging into. The plot test is not included in our `pytest` test suite so I'm fine with that one failing.

**Review by prismofeverything — APPROVED** [review]
Huge improvement all around, thanks for bringing `arrow` into the modern age (!) This pattern is very useful - if it were worthwhile this could be the basis of a cython template that could be reused in the future (just had to go through a similar thing recently).

**Comment by prismofeverything** [comment]
I granted trusted publisher access to this repository for the package on pypi in the `release.yml` workflow, let me know if additional configuration is required there.

**Pull request merged**